Fixed #27448 -- Don't use deprecated PostGIS functions, depending on PostGIS version.

roques opened the pull request:
PostGIS 2.2.0 deprecated `ST_distance_sphere` in favour of `ST_DistanceSphere` and a few other functions. See: https://trac.osgeo.org/postgis/ticket/2748
The old functions are still available, but now issue deprecation warnings, which significantly slows them down (>40× for me).
`django.contrib.db.backends.postgis` currently requires PostGIS ≥2.1.0. Unfortunately, the not deprecated replacement functions only became available with PostGIS 2.2.0. So, to retain compatibility with PostGIS 2.1.0 we have to switch functions depending on the PostGIS version.

This branch contains:
 * A test-case for the deprecated functions being used (Thanks Mjumbe Wawatu Poe for the initial patch!)
 * Some trivial documentation updates.
 * A simple fix, which determines the functions to use in `PostGISOperations.__init__`
 * A test-case for the fix.
 * A more lazily evaluated, but a bit larger fix, which does not confuse Django's testsuite by evaluating an SQL-query when instantiating `PostGISOperations`.

I'll squash it all into one commit after review and patch-up. Either with or without the lazy fix.

Ticket: https://code.djangoproject.com/ticket/27448

timgraham commented:
buildbot, add to whitelist.

claudep commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -163,8 +158,8 @@ def __init__(self, connection):
-        self.distance_sphere = prefix + 'distance_sphere'
-        self.distance_spheroid = prefix + 'distance_spheroid'
+        # self.distance_sphere =  ... see @cached_property below
+        # self.distance_spheroid =  ... see @cached_property below
```
All those `self.function` definitions are used by `GeoQuerySet` which is currently deprecated. I wouldn't touch them and would concentrate only on function names.

roques replied:
I've reverted all fixes related to GeoQuerySet.
Please take a look at the code, as I'm unsure about some of the stylistic choices I made.

claudep commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -91,13 +91,14 @@ def as_sql(self, connection, lookup, template_params, sql_params):
             if len(lookup.rhs) == 3 and lookup.rhs[-1] == 'spheroid':
-                template_params.update({'op': self.op, 'func': 'ST_Distance_Spheroid'})
+                template_params['op'] = self.op
+                template_params['func'] = connection.ops.spatial_function_name("DistanceSpheroid")
```
Keep the `update()` syntax (like below).

claudep commented on tests/gis_tests/distapp/tests.py:
```diff
@@ -645,6 +645,19 @@ def test_distance_order_by(self):
+    @skipUnlessDBFeature("has_Distance_function")
+    def test_distance_not_deprecated(self):
+        if postgis:
+            hillsdale = AustraliaCity.objects.get(name='Hillsdale')
```
As the whole test is postgis-specific, use a decorator instead (`@skipUnless(postgis, "This is a PostGIS-specific test")`

claudep commented on tests/gis_tests/distapp/tests.py:
```diff
@@ -645,6 +645,19 @@ def test_distance_order_by(self):
+        if postgis:
+            hillsdale = AustraliaCity.objects.get(name='Hillsdale')
+            qs = AustraliaCity.objects.annotate(distance=Distance('point', hillsdale.point, spheroid=True))
+            [c.distance.m for c in qs]    # just evaluate
```
Wrapping the above line in `list()` will also force evalutation.

roques replied:
Ah, yes!  ... and having `list(...)` without using the result makes it obvious (to me at least) that the whole purpose of `list()` is to force evaluation, so no comment required, or?

claudep replied:
no comment

roques commented on django/contrib/gis/db/models/functions.py:
```diff
@@ -249,11 +249,11 @@ def as_postgresql(self, compiler, connection):
             # Geometry fields with geodetic (lon/lat) coordinates need special distance functions
             if self.spheroid:
-                self.function = 'ST_Distance_Spheroid'  # More accurate, resource intensive
+                self.function = connection.ops.spatial_function_name('DistanceSpheroid')
```
Flake8 complains about the end-of-line comment making the line too long.
I don't think that particular comment is very helpful anyway. This is not really the right place to learn the difference between `DistanceSphere` and `DistanceSpheroid`. But, as there maybe are people reading the code for which the comment will be helpful, I'll move it to a new line before the assignment.

claudep commented:
Great, could you please merge commits? The commit message should also use past tense (read https://docs.djangoproject.com/en/1.10/internals/contributing/committing-code/#committing-guidelines).
@timgraham Would you mind a quick final review?

roques commented:
I squashed the commits and rewrote the commit message. But, I'm not overly confident that I matched the wanted voice of the message; Is there anything I could still improve?

timgraham commented on django/contrib/gis/db/backends/postgis/operations.py:
```diff
@@ -91,13 +91,16 @@ def as_sql(self, connection, lookup, template_params, sql_params):
-                template_params.update({'op': self.op, 'func': 'ST_Distance_Spheroid'})
+                template_params.update({
+                    'op': self.op,
+                    'func': connection.ops.spatial_function_name("DistanceSpheroid"),
```
Use single quotes for consistency (and below)

timgraham commented on tests/gis_tests/distapp/tests.py:
```diff
@@ -645,6 +647,18 @@ def test_distance_order_by(self):
         ).values_list('name', flat=True).filter(name__in=('San Antonio', 'Pearland'))
         self.assertSequenceEqual(qs, ['San Antonio', 'Pearland'])
 
+    @skipUnless(postgis, "This is a PostGIS-specific test")
```
I think we could skip this test, as it seems a bit fragile and not really needed given the other tests.

timgraham commented on tests/gis_tests/tests.py:
```diff
@@ -94,3 +94,19 @@ def test_no_version_number(self):
+
+    def test_version_dependent_funcs(self):
+        """
+        Test that PostGIS version dependent functions are switched accordingly.
```
The guidelines from "[Coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style)" are: In test docstrings, state the expected behavior that each test demonstrates. Don’t include preambles such as “Tests that” or “Ensures that”.

Could you also add something like "Remove when dropping support for PostGIS 2.1"

roques replied:
I've added the "Remove when dropping support for PostGIS 2.1", but feel that it should be unnecessary. The logical place for this text would be as a comment in `PostGISOperations.function_names()`, but that code is so obviously useless after dropping PostGIS 2.1 support and easily locatable when searching for `spatial_version` which everybody should do when raising the minimum `spatial_version` supported. Furthermore, once `function_names` has been changed back to a regular property to drop PostGIS <2.2 support, `test_version_dependent_funcs` will fail and it will be obvious that it is only needed to test backwards-compatibility with PostGIS <2.2. — But, you know better which hints for future maintenance you appreciate :-)
